kernel/frame_allocator/src/lib.rs

```diff
@@ -382,24 +572,61 @@ impl AllocatedFrames {
             return Err(other);
         }
 
-        // ensure the now-merged AllocatedFrames doesn't run its drop handler and free its frames.
+        // ensure the now-merged Frames doesn't run its drop handler
         core::mem::forget(other); 
         Ok(())
     }
 
-    /// Splits this `AllocatedFrames` into two separate `AllocatedFrames` objects:
+    /// Splits up the given `Frames` into multiple smaller `Frames`.
+    /// 
+    /// Returns a tuple of three `Frames`:
+    /// 1. The `Frames` containing the requested range of frames starting at `start_frame`.
+    /// 2. The range of frames in the `self` that came before the beginning of the requested frame range.
+    /// 3. The range of frames in the `self` that came after the end of the requested frame range.
+    /// 
+    /// If `start_frame` is not contained within `self` or `num_frames` results in an end frame greater than the end of `self`,
+    /// then `self` is not changed and we return (self, None, None).
+    pub fn split(
```

Comment:
Sure, that's a bit long though. I'd say to let the name of the below `SplitRange` struct define the name of the function (based on whatever you end up calling that type).

Reply:
ok so i skimmed the rust std docs for Vec and slice, and on second thought I don't think `extract` is a good choice because it has a bit of a name conflict with `extract_if` (which is the new name of the feature previously known as `drain_filter`).

`split_*` seems like the best, so let's just go with `split_range`. The `Vec` type also has a function called `split_off()`, which we could follow, but in my mind that invokes an idea of just one chunk being split off into two, whereas we have three.

Reply:
agree about split_off, I think split_range is good.